Code size optimizations: Replace .bind() with arrow functions, add listeners in constructors.

justinfagnani opened the pull request:
I didn't make an issue for this, but I saw a few things in the course that could be optimized a bit.

What I did:

- Replaced the `this.handleX = this.handleX.bind(this)` pattern with arrow-function-valued class fields. This is a bit more compact in source, and likely microscopically faster at runtime. `this.x.bind(this)` is still a good pattern if you want to allow the bound method to be overridden in subclasses, but in all of these cases the method was private, so that wasn't possible.
- Moved `this.addEventListener()` calls from `connectedCallback()` to the constructor. Event listeners added on `this` don't need to be removed on disconnection. The browser will collect the cycle of host element and even listener when the element is no longer reachable. This removes a bunch of `removeEventListener()` calls and sometimes `connectedCallback()` and `disconnectedCallback()` completely.
- Moved some small logic (`new Modal`) from connected to constructor.

justinfagnani committed:
Replace .bind() with arrow functions, add listeners in constructors.

justinfagnani committed:
Migrate SlCarousel.handleSlotChange

mpharoah commented:
> Moved this.addEventListener() calls from connectedCallback() to the constructor. Event listeners added on this don't need to be removed on disconnection. The browser will collect the cycle of host element and even listener when the element is no longer reachable. This removes a bunch of removeEventListener() calls and sometimes connectedCallback() and disconnectedCallback() completely.

Is this necessarily true for all standards conforming browsers? The Lit Element documentation itself recommends using `connectedCallback` and `disconnectedCallback` for managing event listeners.
> In `connectedCallback()` you should setup tasks that should only occur when the element is connected to the document. The most common of these is adding event listeners to nodes external to the element, like a keydown event handler added to the window. Typically, anything done in `connectedCallback()` should be undone when the element is disconnected — for example, removing event listeners on window to prevent memory leaks.

KonnorRogers commented on src/components/tab-group/tab-group.ts:
```diff
@@ -70,7 +70,7 @@ export default class SlTabGroup extends ShoelaceElement {
   @property({ attribute: 'no-scroll-controls', type: Boolean }) noScrollControls = false;
 
   connectedCallback() {
-    const whenAllDefined = Promise.allSettled([
```
Hmm....Any reason for changing this from `Promise.allSettled` to `Promise.all`? 

As I understand `Promise.all` could early error and then never wait for the 2nd element to be defined.

justinfagnani replied:
`whenDefined()` won't reject, so there's no reason to use `allSettled()`. Even if it did reject, you wouldn't want to continue on errors here. `all()` is shorter and doesn't create the unused result objects.

KonnorRogers replied:
Thats fair...didnt think it was a big deal, just making sure I didn't miss anything!

KonnorRogers commented on src/components/tooltip/tooltip.ts:
```diff
@@ -98,23 +98,20 @@ export default class SlTooltip extends ShoelaceElement {
 
+  constructor() {
+    super();
+    // TODO (justinfagnani): does this need to be done in firstUpdated for some
```
🤔 perhaps the handlers need the inner shadow dom to be rendered? In which case they should probably just have guard clauses.

cc: @claviska any insight here?

KonnorRogers reviewed (approved):
Looks fine to me. static class properties are supported as of Safari 14.5 so I think its fine on the support matrix. 

https://caniuse.com/mdn-javascript_classes_static_class_fields

One question on `Promise.all` vs `Promise.allSettled`. Otherwise I see no issue with this. I'm curious on actual bundle size / perf changes, but thats a future problem, definitely wanna get perf + bundle size benchmarks for the future.

justinfagnani commented:
@KonnorRogers 

> static class properties are supported as of Safari 14.5 so I think its fine on the support matrix.

Isn't that controlled by the target in tsconfig and not dependent on what language features are used in source?

KonnorRogers commented:
Controlled by target in ESBuild, but same thing. You're right @justinfagnani 🤦‍♂️ .

claviska commented:
> The browser will collect the cycle of host element and even listener when the element is no longer reachable. This removes a bunch of removeEventListener() calls and sometimes connectedCallback() and disconnectedCallback() completely.

@justinfagnani Just to be sure, what happens if we:

1. Append an element to the DOM and store a reference in `el`
2. `connectedCallback` runs and attaches listeners
3. We call `el.remove()` but we maintain the reference
4. After an arbitrary amount of time, we append `el` to the DOM again
5. `connectedCallback` runs again and attaches listeners again

Will we have duplicate listeners now, or is removing the element sufficient for all browsers to garbage collect the listeners? Will garbage collection be fast enough and/or guaranteed even if the element is removed and very quickly re-appended? Do browsers act consistently here?

It makes me a bit nervous relying on internal browser behaviors that aren't clearly documented and don't necessarily work the same way in Blink, Webkit, Gecko, et al.

justinfagnani commented:
@claviska I moved most of the `addEventListener()` calls to the constructor, which only runs once. So there's no deduping or need for it in those cases. Any event listeners added to another object like document or window were left as-is in `connectedCallback()` and cleaned up in `disconnectedCallback()`. Those _do_ need to be cleaned up to prevent leaks.

This isn't relying on undocumented browser behavior - cleaning up cycles is the intended behavior, probably specc'ed somewhere, and just broken in very old IE version (pre IE 7, I believe).
